File: Python_pyxel/Collection.py
import time

def gcd(n1, n2):
    '''
    最大公約数を見つけます

    条件

    n1 >= n2
    '''

    if n1 % n2 == 0:
        return n2
    else:
        return gcd(n2, n1 % n2)

class prime():
    '''
    primelist にn以下の素数を格納します

    coprimelist にn以下のnと互いに素な数を格納します
    '''

    primelist = []
    coprimelist = []
    n_prime = 0

    def __init__(self, n = 0):
        self.primelist = prime.primenumberlist(n)
        self.n_prime = prime.theprimenumber(n)
        self.coprimelist = prime.coprimenumberlist(n)

    def primenumberlist(n):
        '''
        n 以下の素数をリストで返します
        '''

        n_list = [x for x in range(5, n + 1, 2) if x % 3 != 0]
        prime_list = [2, 3]
        for _ in range(n):
            if prime_list[-1] ** 2 >= n:
                prime_list += n_list
                break
            else:
                prime_list.append(n_list[0])
                n_list = [x for x in n_list if x % prime_list[-1] != 0]
            if len(n_list) == 0:
                break
        return prime_list
    
    def theprimenumber(n):
        '''
        n 番目の素数を返します
        '''

        prime_list = [2, 3]
        i = 1
        while len(prime_list) < n:
            for j in range(2):
                check = True
                for s in prime_list:
                    if (6 * i + (-1) ** (j + 1)) % s == 0:
                        check = False
                        break
                    if s ** 2 >= 6 * i + (-1) ** (j + 1):
                        break
                if check:
                    prime_list.append(6 * i + (-1) ** (j + 1))
            i += 1
        return prime_list[n - 1]

    def coprimenumberlist(n):
        '''
        n と互いに素な n 未満の数をリストで返します
        '''

        cpri = [1]
        for i in range(2, n):
            if gcd(n, i) == 1:
                cpri.append(i)
        return cpri
    
    def primefactorization(n):
        '''
        n の素因数をリストで返します
        '''

        primefac = []
        m = n
        for i in range(2, n + 1):
            while m % i == 0:
                primefac.append(i)
                m /= i
        return primefac
    
    def isprime(n):
        '''
        n が素数かどうかを調べ bool で返します
        '''

        i = 1
        for j in range(2, 4):
            if n % j == 0:
                return False
        
        while True:
            for j in range(1, 3):
                nearprime = 6 * i + (-1) ** j
                if nearprime ** 2 > n:
                    return True
                elif n % nearprime == 0:
                    return False

class LCG():
    '''
    線形合同法 (Linear congruential generators) による乱数を生成します
    '''

    __seed = 0
    __a = 0
    __c = 0
    __m = 1
    __x = 0
    __start = 0

    def __init__(self, start, end):
        if start >= end:
            #　start が end 以上の場合、計算で起こる無限ループ回避のため初期設定をしない
            print("randomliner : start < end")
        else:
            #　初期設定をする
            #　a, c, m, seed の値を取得
            self.__start = start
            self.__m = end
            self.__a, self.__c = self.__find_a_c()
            self.__seed = pow(int(time.time() * 1000), 1, self.__m)
    
    def  __find_a_c(self):
        '''
        a と c の値を計算して return します
        '''

        #　a-1 の値を求める
        a = 4
        j = 0
        for i in prime.primefactorization(self.__m):
            if (j != i):
                a *= i
                j = i
        
        #　c の値を求める
        n_coprime = prime(self.__m).coprimelist
        c = n_coprime[len(n_coprime) // 2]
        
        return a + 1, c

    def random(self):
        '''
        生成した乱数を返します
        '''

        #　seed値が 0 以上か、生成した数字が start 未満のとき繰り返し計算する
        self.__x = (self.__a * self.__x + self.__c) % self.__m
        while self.__seed > 0 or self.__x < self.__start:
            self.__x = (self.__a * self.__x + self.__c) % self.__m
            self.__seed -= 1
        return self.__x

class ReLCG():
    '''
    LCGを改良した乱数を生成します
    '''

    __seed = 0
    __a = 0
    __c = 0
    __m = 10000
    __x = 0
    __start = 0
    __end = 0 #　追加

    def __init__(self, start, end):
        if start >= end:
            #　start が end 以上の場合、計算で起こる無限ループ回避のため初期設定をしない
            print("randomliner : start < end")
        else:
            #　初期設定をする
            #　a, c, m, seed の値を取得
            self.__start = start
            self.__end = end #　追加
            if self.__m < end: #　追加
                self.__m = end
            
            self.__a, self.__c = self.__find_a_c()
            self.__seed = pow(int(time.time() * 1000), 1, self.__m)
    
    def  __find_a_c(self):
        '''
        a と c の値を計算して return します
        '''

        #　a-1 の値を求める
        a = 4
        j = 0
        for i in prime.primefactorization(self.__m):
            if (j != i):
                a *= i
                j = i
        
        #　c の値を求める
        n_coprime = prime(self.__m).coprimelist
        c = n_coprime[len(n_coprime) // 2]
        
        return a + 1, c

    def random(self):
        '''
        生成した乱数を返します
        '''

        #　seed値が 0 以上か、生成した数字が start 未満のとき繰り返し計算する
        self.__x = (self.__a * self.__x + self.__c) % self.__m
        generated = self.__x * (self.__end - self.__start) // self.__m + self.__start #　追加
        while self.__seed > 0: #　一部変更
            self.__x = (self.__a * self.__x + self.__c) % self.__m
            self.__seed -= 1
        return generated #　変更

File: Csharp_unity_3D/Assets/Scripts/Collection.cs
using System;
using System.Collections;
using System.Collections.Generic;
using System.Linq;
using UnityEngine;

namespace Collection
{
    //-----------------------#1-------------------------------
    public class Calculate
    {
        /// <summary>
        /// 二つの数字の最大公約数を求めます
        /// </summary>
        /// <param name="n1"></param>
        /// <param name="n2"></param>
        /// <returns></returns>
        public int GCD(int n1, int n2)
        {
            if (n1 % n2 == 0)
            {
                return n2;
            }
            else
            {
                return GCD(n2, n1 % n2);
            }
        }
    }

    //-----------------------#2-------------------------------

    /// <summary>
    /// 素な数に関するクラス
    /// </summary>
    public class Prime
    {
        public int[] primelist = {};
        public int theprime = 0;
        public int[] coprimelist = {};
        public Prime(int n)
        {
            primelist = PrimeNumberList(n);
            theprime = ThePrimeNumber(n);
            coprimelist = CoPrimeNumberList(n);
        }

        /// <summary>
        /// n以下の素数を返します
        /// </summary>
        /// <param name="n"></param>
        /// <returns></returns>
        public static int[] PrimeNumberList(int n)
        {
            List<int> prime_list = new List<int> {2, 3};
            int[] n_list = Enumerable.Range(0, (n - 5) / 2 + 1).Select(x => 5 + x * 2).Where(x => x % 3 != 0).ToArray();

            for (int i = 0; i < n; i++)
            {
                if (Math.Pow(prime_list[prime_list.Count - 1], 2) >= n)
                {
                    prime_list.AddRange(n_list.ToList());
                    break;
                }
                else
                {
                    prime_list.Add(n_list[0]);
                    n_list = n_list.Where(x => x % n_list[0] != 0).ToArray();
                }
                if (n_list.Length == 0)
                {
                    break;
                }
            }

            return prime_list.ToArray();
        }

        /// <summary>
        /// n番目の素数を返します
        /// </summary>
        /// <param name="n"></param>
        /// <returns></returns>
        public static int ThePrimeNumber(int n)
        {
            List<int> primelist = new List<int> {2, 3};
            int i = 1;
            while (primelist.Count < n)
            {
                for (int j = 0; j < 2; j++)
                {
                    bool check = true;
                    foreach (int s in primelist)
                    {
                        if ((6 * i + (int)Math.Pow(-1, j + 1)) % s == 0)
                        {
                            check = false;
                            break;
                        }
                        if (Math.Pow(s, 2) >= 6 * i + (int)Math.Pow(-1, j + 1))
                        {
                            break;
                        }
                    }
                    if (check)
                    {
                        primelist.Add(6 * i + (int)Math.Pow(-1, j + 1));
                    }
                }
                i++;
            }
            return primelist[n - 1];
        }

        /// <summary>
        /// n以下のnと互いに素な数を返します
        /// </summary>
        /// <param name="n"></param>
        /// <returns></returns>
        public static int[] CoPrimeNumberList(int n)
        {
            List<int> coprimelist = new List<int> {1};
            Calculate cal = new Calculate();
            for (int i = 2; i < n; i++)
            {
                if (cal.GCD(n, i) == 1)
                {
                    coprimelist.Add(i);
                }
            }
            return coprimelist.ToArray();
        }

        /// <summary>
        /// nを素因数分解します
        /// </summary>
        /// <param name="n"></param>
        /// <returns></returns>
        public static int[] PrimeFactorization(int n)
        {
            List<int> primefac = new List<int> {};
            int m = n;
            for (int i = 2; i < n + 1; i++)
            {
                while (m % i == 0)
                {
                    primefac.Add(i);
                    m /= i;
                }
                if (m == 1)
                {
                    break;
                }
            }
            return primefac.ToArray();
        }

        /// <summary>
        /// n が素数かどうかを判定します
        /// </summary>
        /// <param name="n"></param>
        /// <returns></returns>
        public static bool IsPrime(int n)
        {
            int i = 1;
            for (int j = 2; j < 4; j++)
            {
                if (n % j == 0)
                {
                    return false;
                }
            }

            while (true)
            {
                for (int j = 1; j < 3; j++)
                {
                    int nearprime = 6 * i + (int)Math.Pow(-1, j);
                    if (Math.Pow(nearprime, 2) > n)
                    {
                        return true;
                    }
                    else if (n % nearprime == 0)
                    {
                        return false;
                    }
                }
            }
        }
    }

    //-----------------------#3-------------------------------

    /// <summary>
    ///　線形合同法 (Linear congruential generators) による乱数を生成します
    /// </summary>
    class LCG
    {
        private int seed;
        private int a;
        private int c;
        private int m;
        private int x = 0;
        private int start;

        /// <summary>
        /// start 以上 end 未満の乱数を生成します
        /// </summary>
        /// <param name="start"></param>
        /// <param name="end"></param>
        public LCG(int start, int end)
        {
            if (start >= end)
            {
                //　start が end 以上の場合、計算で起こる無限ループ回避のため初期設定をしない
                Debug.Log("randomliner : start < end");
            }
            else
            {
                //　初期設定をする
                //　a, c, m, seed の値を取得
                this.start = start;
                this.seed = DateTime.Now.Millisecond % end;
                this.m = end;
                int[] a_c = find_a_c(end);
                this.a = a_c[0];
                this.c = a_c[1];
            }
        }

        /// <summary>
        /// a と c の値を求めます
        /// </summary>
        /// <param name="m"></param>
        /// <returns></returns>
        private int[] find_a_c(int m)
        {
            //　a_c を定義、0番目が a、 1番目が c
            int[] a_c = {4, 0};

            //　a の値を求める
            int j = 0;
            foreach (int i in Prime.PrimeFactorization(m))
            {
                if (i != j)
                {
                    a_c[0] *= i;
                    j = i;
                }
            }
            a_c[0] += 1;

            //　c の値を求める
            int[] coli = Prime.CoPrimeNumberList(m);
            a_c[1] = coli[coli.Length / 2];
            return a_c;
        }

        /// <summary>
        ///　生成した乱数を返します
        /// </summary>
        /// <returns></returns>
        public int random()
        {
            //　乱数を生成、生成した値が start 未満または seed が 1以上なら再生成
            this.x = (this.a * this.x + this.c) % this.m;
            while (this.seed > 0 || this.start > this.x)
            {
                this.x = (this.a * this.x + this.c) % this.m;
                this.seed -= 1;
            }
            return this.x;
        }
    }

    /// <summary>
    ///　LCGを改良した乱数を生成します
    /// </summary>
    class ReLCG
    {
        private int seed;
        private int a;
        private int c;
        private int m = 10000;
        private int x = 0;
        private int start;
        private int end; //　追加

        /// <summary>
        /// start 以上 end 未満の乱数を生成します
        /// </summary>
        /// <param name="start"></param>
        /// <param name="end"></param>
        public ReLCG(int start, int end)
        {
            if (start >= end)
            {
                //　start が end 以上の場合、計算で起こる無限ループ回避のため初期設定をしない
                Debug.Log("randomliner : start < end");
            }
            else
            {
                //　初期設定をする
                //　a, c, m, seed の値を取得
                this.start = start;
                this.seed = DateTime.Now.Millisecond % this.m;
                this.end = end; //　追加
                if (this.m < end) //　追加
                {
                    this.m = end;
                }
                int[] a_c = find_a_c(end);
                this.a = a_c[0];
                this.c = a_c[1];
            }
        }

        /// <summary>
        /// a と c の値を求めます
        /// </summary>
        /// <param name="m"></param>
        /// <returns></returns>
        private int[] find_a_c(int m)
        {
            //　a_c を定義、0番目が a、 1番目が c
            int[] a_c = {4, 0};

            //　a の値を求める
            int j = 0;
            foreach (int i in Prime.PrimeFactorization(m))
            {
                if (i != j)
                {
                    a_c[0] *= i;
                    j = i;
                }
            }
            a_c[0] += 1;

            //　c の値を求める
            int[] coli = Prime.CoPrimeNumberList(m);
            a_c[1] = coli[coli.Length / 2];
            return a_c;
        }

        /// <summary>
        ///　生成した乱数を返します
        /// </summary>
        /// <returns></returns>
        public int random()
        {
            //　乱数を生成、生成した値が start 未満または seed が 1以上なら再生成
            this.x = (this.a * this.x + this.c) % this.m;
            int generated = this.x * (this.end - this.start) / this.m + this.start; //　追加
            while (this.seed > 0 || generated < this.start) //　一部変更
            {
                this.x = (this.a * this.x + this.c) % this.m;
                this.seed -= 1;
                generated = this.x * (this.end - this.start) / this.m + this.start; //　追加
            }
            return generated; //　変更
        }
    }
}
